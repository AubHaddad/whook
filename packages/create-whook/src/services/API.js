import { name, autoService } from 'knifecycle';

import { definition as getPingDefinition } from '../handlers/getPing';
import { definition as getTimeDefinition } from '../handlers/getTime';

export default name('API', autoService(initAPI));

async function initAPI({ CONFIG, log }) {
  log('debug', 'ðŸ¦„ - Initializing the API service!');

  const API = {
    host: CONFIG.host,
    basePath: CONFIG.basePath,
    schemes: CONFIG.schemes,
    securityDefinitions: {
      basicAuth: {
        type: 'basic',
      },
    },
    swagger: '2.0',
    info: {
      version: CONFIG.version,
      title: CONFIG.name,
      description: CONFIG.description,
    },
    paths: {
      [getPingDefinition.path]: {
        [getPingDefinition.method]: getPingDefinition.operation,
      },
      [getTimeDefinition.path]: {
        [getTimeDefinition.method]: getTimeDefinition.operation,
      },
    },
  };

  return API;
}
