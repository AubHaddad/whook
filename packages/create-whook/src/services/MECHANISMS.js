import HTTPError from 'yhttperror';
import { name, autoService } from 'knifecycle';
import { BEARER as BEARER_MECHANISM } from 'http-auth-utils';

export const FAKE_MECHANISM = {
  type: 'Fake',
  parseAuthorizationRest: rest => {
    let userId;
    let scopes;

    rest.replace(/^(\d+)-((\w+,)*(\w+){1})$/, (_, rawUserId, rawScopes) => {
      userId = parseInt(rawUserId);
      scopes = rawScopes.split();
      return '';
    });

    if ('undefined' === typeof userId || 'undefined' === typeof scopes) {
      throw new HTTPError(400, 'E_INVALID_FAKE_TOKEN');
    }

    return {
      hash: rest,
      userId,
      scopes,
    };
  },
};

export default name('MECHANISMS', autoService(initMechanisms));

async function initMechanisms({ DEBUG_NODE_ENVS, NODE_ENV, log }) {
  log('debug', 'üîß - Initializing auth mechanisms');

  const debugging = DEBUG_NODE_ENVS.includes(NODE_ENV);
  const MECHANISMS = [BEARER_MECHANISM, ...(debugging ? [FAKE_MECHANISM] : [])];

  if (debugging) {
    log('warning', '‚ö†Ô∏è - Using fake auth mechanism!');
  }
  return MECHANISMS;
}
