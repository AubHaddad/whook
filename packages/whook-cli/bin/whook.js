const path = require('path');
const run = require('../dist/index').default;
const { prepareEnvironment } = require(path.join(
  process.cwd(),
  'dist',
  'index',
));

run({ prepareEnvironment });
