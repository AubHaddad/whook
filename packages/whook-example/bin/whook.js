#! /usr/bin/env node

import path from 'path';
import runCLI from '@whook/whook';

const { prepareEnvironment } = await import(
  path.join(process.cwd(), 'src', 'index.js')
);

await runCLI(prepareEnvironment);
