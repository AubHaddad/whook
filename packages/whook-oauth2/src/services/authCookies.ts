import ms from 'ms';
import cookie from 'cookie';
import { autoService } from 'knifecycle';
import type { BaseAuthenticationData } from '@whook/authorization';

export const AUTH_API_PREFIX = '/auth';

export type AuthHandlersConfig<
  AUTHENTICATION_DATA extends BaseAuthenticationData = BaseAuthenticationData,
> = {
  ROOT_AUTHENTICATION_DATA: AUTHENTICATION_DATA;
};

export type AuthCookiesConfig = {
  ENV?: {
    DEV_MODE?: string;
  };
  COOKIES: {
    domain: string;
  };
  BASE_PATH?: string;
};

export type AuthCookiesData = {
  refresh_token: string;
  access_token: string;
};

export type AuthCookiesService = {
  build: (
    data?: Partial<AuthCookiesData>,
    options?: { session: boolean },
  ) => string[];
  parse: (content: string) => Partial<AuthCookiesData>;
};

export default autoService(initAuthCookies);

async function initAuthCookies({
  ENV = {},
  COOKIES,
  BASE_PATH = '',
}: AuthCookiesConfig): Promise<AuthCookiesService> {
  function build(data: Partial<AuthCookiesData> = {}, { session = true } = {}) {
    return [
      cookie.serialize('access_token', data.access_token || '', {
        path: BASE_PATH + AUTH_API_PREFIX,
        httpOnly: true,
        domain: ENV.DEV_MODE ? undefined : COOKIES.domain,
        secure: !ENV.DEV_MODE,
        ...(ENV.DEV_MODE ? {} : { sameSite: true }),
        ...(data.access_token ? {} : { maxAge: 0 }),
      }),
      cookie.serialize('refresh_token', data.refresh_token || '', {
        path: BASE_PATH + AUTH_API_PREFIX,
        httpOnly: true,
        domain: ENV.DEV_MODE ? undefined : COOKIES.domain,
        secure: !ENV.DEV_MODE,
        ...(ENV.DEV_MODE ? {} : { sameSite: true }),
        ...(session ? {} : { maxAge: Math.round(ms('100y') / 1000) }),
      }),
    ];
  }

  function parse(cookieHeader: string): Partial<AuthCookiesData> {
    const data = cookie.parse(cookieHeader);

    if (data.access_token && data.refresh_token) {
      return {
        access_token: data.access_token,
        refresh_token: data.refresh_token,
      };
    }

    return {};
  }

  return {
    build,
    parse,
  };
}
