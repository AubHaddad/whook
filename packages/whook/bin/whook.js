#! /usr/bin/env node

import path from 'path';
import runCLI from '../dist/cli.js';

const { prepareEnvironment } = await import(
  path.join(process.cwd(), 'dist', 'index.js')
);

await runCLI(prepareEnvironment);
