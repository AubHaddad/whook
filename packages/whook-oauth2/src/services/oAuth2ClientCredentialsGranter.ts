import { autoService } from 'knifecycle';
import { LogService } from 'common-services';
import { noop } from '@whook/whook';
import YError from 'yerror';
import {
  OAuth2GranterService,
  CheckApplicationService,
} from './oAuth2Granters';
import { BaseAuthenticationData } from '@whook/authorization';

export type OAuth2ClientCredentialsGranterDependencies = {
  checkApplication: CheckApplicationService;
  log?: LogService;
};
export type OAuth2ClientCredentialsGranterParameters<
  AUTHENTICATION_DATA extends BaseAuthenticationData = BaseAuthenticationData
> = {
  username: string;
  password: string;
  scope?: AUTHENTICATION_DATA['scope'];
};
export type OAuth2ClientCredentialsGranterService<
  AUTHENTICATION_DATA extends BaseAuthenticationData = BaseAuthenticationData
> = OAuth2GranterService<
  unknown,
  unknown,
  OAuth2ClientCredentialsGranterParameters,
  AUTHENTICATION_DATA
>;

export default autoService(initOAuth2ClientCredentialsGranter);

// Client Credentials Grant
// https://tools.ietf.org/html/rfc6749#section-4.4
async function initOAuth2ClientCredentialsGranter({
  checkApplication,
  log = noop,
}: OAuth2ClientCredentialsGranterDependencies): Promise<
  OAuth2ClientCredentialsGranterService
> {
  const authenticateWithClientCredentials: OAuth2ClientCredentialsGranterService['authenticator']['authenticate'] = async (
    { scope: demandedScope = '' },
    authenticationData,
  ) => {
    // The client must be authenticated
    if (!authenticationData) {
      throw new YError('E_UNAUTHORIZED');
    }

    // Checking the scope and availability of the demanded
    // grant type
    await checkApplication({
      applicationId: authenticationData.applicationId,
      type: 'client_credentials',
      scope: demandedScope,
    });

    return {
      ...authenticationData,
      scope: demandedScope,
    };
  };

  log('debug', 'ðŸ‘« - OAuth2ClientCredentialsGranter Service Initialized!');

  return {
    type: 'client_credentials',
    authenticator: {
      grantType: 'client_credentials',
      authenticate: authenticateWithClientCredentials,
    },
  };
}
