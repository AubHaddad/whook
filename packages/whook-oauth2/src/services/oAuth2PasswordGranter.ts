import { autoService } from 'knifecycle';
import { LogService } from 'common-services';
import { noop } from '@whook/whook';
import YError from 'yerror';
import {
  OAuth2GranterService,
  OAuth2PasswordService,
  CheckApplicationService,
} from './oAuth2Granters';
import { BaseAuthenticationData } from '@whook/authorization';

export type OAuth2PasswordGranterDependencies = {
  oAuth2Password: OAuth2PasswordService;
  checkApplication: CheckApplicationService;
  log?: LogService;
};
export type OAuth2PasswordGranterParameters = {
  username: string;
  password: string;
  scope?: string;
};
export type OAuth2PasswordGranterService<
  AUTHENTICATION_DATA extends BaseAuthenticationData = BaseAuthenticationData
> = OAuth2GranterService<
  unknown,
  unknown,
  OAuth2PasswordGranterParameters,
  AUTHENTICATION_DATA
>;

export default autoService(initOAuth2PasswordGranter);

// Resource Owner Password Credentials Grant
// https://tools.ietf.org/html/rfc6749#section-4.3
async function initOAuth2PasswordGranter({
  oAuth2Password,
  checkApplication,
  log = noop,
}: OAuth2PasswordGranterDependencies): Promise<OAuth2PasswordGranterService> {
  const authenticateWithPassword: OAuth2PasswordGranterService['authenticator']['authenticate'] = async (
    { username, password, scope: demandedScope = '' },
    authenticationData,
  ) => {
    // The client must be authenticated
    if (!authenticationData) {
      throw new YError('E_UNAUTHORIZED');
    }

    await checkApplication({
      applicationId: authenticationData.applicationId,
      type: 'password',
      scope: demandedScope,
    });

    const finalAuthenticationData = await oAuth2Password.check(
      authenticationData,
      username,
      password,
      demandedScope,
    );

    return finalAuthenticationData;
  };

  log('debug', 'ðŸ‘« - OAuth2PasswordGranter Service Initialized!');

  return {
    type: 'password',
    authenticator: {
      grantType: 'password',
      authenticate: authenticateWithPassword,
    },
  };
}
