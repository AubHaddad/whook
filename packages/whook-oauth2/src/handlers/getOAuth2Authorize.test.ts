import initGetOAuth2Authorize from './getOAuth2Authorize';
import { DEFAULT_ERRORS_DESCRIPTORS } from '@whook/whook';
import { OAUTH2_ERRORS_DESCRIPTORS } from '../services/oAuth2Granters';
import YError from 'yerror';
import { OAuth2Options, OAuth2GranterService } from '..';

describe('getOAuth2Authorize', () => {
  const OAUTH2: OAuth2Options = {
    authenticateURL: 'https://auth.example.com/sign_in',
  };
  const ERRORS_DESCRIPTORS = {
    ...DEFAULT_ERRORS_DESCRIPTORS,
    ...OAUTH2_ERRORS_DESCRIPTORS,
  };
  const log = jest.fn();
  const codeGranter = {
    type: 'code',
    authorizer: { responseType: 'code', authorize: jest.fn() },
    acknowledger: { acknowledgmentType: 'code', acknowledge: jest.fn() },
    authenticator: {
      grantType: 'authorization_code',
      authenticate: jest.fn(),
    },
  };
  const tokenGranter = {
    type: 'token',
    authorizer: { responseType: 'token', authorize: jest.fn() },
    acknowledger: { acknowledgmentType: 'token', acknowledge: jest.fn() },
  };
  const oAuth2Granters: OAuth2GranterService[] = [codeGranter, tokenGranter];

  beforeEach(() => {
    log.mockReset();
    [
      codeGranter.authorizer.authorize,
      codeGranter.acknowledger.acknowledge,
      codeGranter.authenticator.authenticate,
      tokenGranter.authorizer.authorize,
      tokenGranter.acknowledger.acknowledge,
    ].forEach(mock => mock.mockReset());
  });

  test('should redirect', async () => {
    [
      codeGranter.acknowledger.acknowledge,
      codeGranter.authenticator.authenticate,
      tokenGranter.authorizer.authorize,
      tokenGranter.acknowledger.acknowledge,
    ].forEach(mock =>
      mock.mockRejectedValueOnce(new YError('E_NOT_SUPPOSED_TO_BE_HERE')),
    );
    codeGranter.authorizer.authorize.mockResolvedValueOnce({
      applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
      redirectURI: 'http://lol',
      scope: 'user',
    });

    const getOAuth2Authorize = await initGetOAuth2Authorize({
      OAUTH2,
      ERRORS_DESCRIPTORS,
      oAuth2Granters,
      log,
    });
    const response = await getOAuth2Authorize({
      response_type: 'code',
      client_id: 'abbacaca-abba-caca-abba-cacaabbacaca',
      redirect_uri: 'https://www.example.com',
      scope: 'user',
      state: 'bancal',
    });

    expect(response).toMatchInlineSnapshot(`
      Object {
        "headers": Object {
          "location": "https://auth.example.com/sign_in?type=code&redirect_uri=http%3A%2F%2Flol&scope=user&client_id=abbacaca-abba-caca-abba-cacaabbacaca&state=bancal",
        },
        "status": 302,
      }
    `);
    expect({
      logCalls: log.mock.calls.filter(args => args[0].endsWith('stack')),
      codeGranterAuthorizerAuthorizeCalls:
        codeGranter.authorizer.authorize.mock.calls,
      codeGranterAcknowledgerAcknowledgeCalls:
        codeGranter.acknowledger.acknowledge.mock.calls,
      codeGranterAuthenticatorAuthenticateCalls:
        codeGranter.authenticator.authenticate.mock.calls,
      tokenGranterAuthorizerAuthorizeCalls:
        tokenGranter.authorizer.authorize.mock.calls,
      tokenGranterAcknowledgerAcknowledgeCalls:
        tokenGranter.acknowledger.acknowledge.mock.calls,
    }).toMatchSnapshot();
  });

  test('should redirect with an error when some', async () => {
    [
      codeGranter.authorizer.authorize,
      codeGranter.acknowledger.acknowledge,
      codeGranter.authenticator.authenticate,
      tokenGranter.authorizer.authorize,
      tokenGranter.acknowledger.acknowledge,
    ].forEach(mock =>
      mock.mockRejectedValueOnce(new YError('E_NOT_SUPPOSED_TO_BE_HERE')),
    );

    const getOAuth2Authorize = await initGetOAuth2Authorize({
      OAUTH2,
      ERRORS_DESCRIPTORS,
      oAuth2Granters,
      log,
    });
    const response = await getOAuth2Authorize({
      response_type: 'yolo',
      client_id: 'abbacaca-abba-caca-abba-cacaabbacaca',
      redirect_uri: 'https://www.example.com',
      scope: 'user',
      state: 'bancal',
    });

    expect(response).toMatchInlineSnapshot(`
      Object {
        "headers": Object {
          "location": "https://auth.example.com/sign_in?redirect_uri=https%3A%2F%2Fwww.example.com&error=unsupported_response_type&error_decription=The+type+%22yolo%22+is+not+supported.&state=bancal",
        },
        "status": 302,
      }
    `);
    expect({
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      codeGranterAuthorizerAuthorizeCalls:
        codeGranter.authorizer.authorize.mock.calls,
      codeGranterAcknowledgerAcknowledgeCalls:
        codeGranter.acknowledger.acknowledge.mock.calls,
      codeGranterAuthenticatorAuthenticateCalls:
        codeGranter.authenticator.authenticate.mock.calls,
      tokenGranterAuthorizerAuthorizeCalls:
        tokenGranter.authorizer.authorize.mock.calls,
      tokenGranterAcknowledgerAcknowledgeCalls:
        tokenGranter.acknowledger.acknowledge.mock.calls,
    }).toMatchSnapshot();
  });
});
