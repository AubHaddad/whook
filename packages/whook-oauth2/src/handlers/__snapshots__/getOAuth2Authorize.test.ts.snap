// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOAuth2Authorize should redirect 2`] = `
Object {
  "codeGranterAcknowledgerAcknowledgeCalls": Array [],
  "codeGranterAuthenticatorAuthenticateCalls": Array [],
  "codeGranterAuthorizerAuthorizeCalls": Array [
    Array [
      Object {
        "clientId": "abbacaca-abba-caca-abba-cacaabbacaca",
        "redirectURI": "https://www.example.com",
        "scope": "user",
      },
      Object {},
    ],
  ],
  "logCalls": Array [],
  "tokenGranterAcknowledgerAcknowledgeCalls": Array [],
  "tokenGranterAuthorizerAuthorizeCalls": Array [],
}
`;

exports[`getOAuth2Authorize should redirect with an error when some 2`] = `
Object {
  "codeGranterAcknowledgerAcknowledgeCalls": Array [],
  "codeGranterAuthenticatorAuthenticateCalls": Array [],
  "codeGranterAuthorizerAuthorizeCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ‘« - OAuth2 initialization error",
      "E_UNKNOWN_AUTHORIZER_TYPE",
    ],
  ],
  "tokenGranterAcknowledgerAcknowledgeCalls": Array [],
  "tokenGranterAuthorizerAuthorizeCalls": Array [],
}
`;
